### DISCUSSION TITLE
Tkinter modern dialogue box suddenly not working

### DISCUSSION THREAD

**Discussion by aliencaocao** [pyinstaller/pyinstaller#6366] [Help] [answered]
I use the following code to show dialogue boxes:

```
from tkinter.ttk import *
from tkinter import messagebox
messagebox.showerror(title='xx', message='xxx')
```

And I use PyInstaller to compile into exe. Today I found out that the modern themed dialogue box is no longer there. I downloaded the exe binary built a month ago, the dialogue box are modern themed, however, the one I just built are no longer. The code relevant to the box are exactly the same, and I am running on exactly the same system. I do not understand where the issue could be.

Here is an example of modern and old:

Old: ![image](https://user-images.githubusercontent.com/20109683/141665837-e99978e2-b808-42bf-9987-a8cf61fc609e.png)

New: ![image](https://user-images.githubusercontent.com/20109683/141665838-4f44d839-7510-4e04-90e7-6949c381fa9c.png)

As you can see, the buttons are completely different.

Both binaries are built with Python 3.9 and latest development build of PyInstaller

From StackOverflow, they are talking about this manifest file that Pyinstaller uses. I have never specifically set that before, and I'm not sure if now I have to manually set it when building to get back the modern dialogue box.

**Comment by rokm** [comment, marked as answer]
> I have never specifically set that before, and I'm not sure if now I have to manually set it when building to get back the modern dialogue box.

You should not have; if there was change in behavior, it is a regression/bug and we will need to address it.

We did make some manifest-related changes (#6203 and #6243 on `develop`, #6248 on `v4` branch) that *could* be the cause. However, I cannot reproduce the problem here - your example code generates an error dialog with "modern" error icon (i.e., unlike the one in your "old" example).

What kind of build are you using? Onefile or onedir (or it doesn't matter)? Windowed or console (or it doesn't matter)? What is the target OS version? And, does this happen only with `develop` branch (assuming "latest development build of PyInstaller" means you are using `develop` branch) or also `v4` branch? Judging by the PR timeline, v4.5.1 release should still work OK, but 4.6 should not?

**Reply by aliencaocao** [reply]
Using the 5.0dev0 that was downloaded about 2 weeks ago.
I just opened a new file and compiled using `pyinstaller -F temp.py`
Using -F or -D does not matter. Target OS is Windows 10. Console or windowed doesn't matter.

Contents:
```
from tkinter import messagebox
messagebox.showinfo(title='xx', message='xxx')
```

And it is giving me the old text box.
![image](https://user-images.githubusercontent.com/20109683/141679190-b7ce49b6-eb74-425d-a5f8-7981439ecb1b.png)
Installed 4.5.1 still did not work. Neither does 4.7.

The 'working' example was built on 3 Oct using 5.0dev0 downloaded on 11 Sep. All build settings are exacting same (because I ran a build script and that script has not changed since a few months ago, so all the command switches are same)

**Reply by rokm** [reply]
> The 'working' example was built on 3 Oct using 5.0dev0 downloaded on 11 Sep.

What if you use `develop` at commit that was at HEAD on 11 Sep (e.g., 708f5d90ff5fc91e1764ab082322ca7faf4d9c7b)? Does that work?

**Reply by aliencaocao** [reply]
> > The 'working' example was built on 3 Oct using 5.0dev0 downloaded on 11 Sep.
> 
> What if you use `develop` at commit that was at HEAD on 11 Sep (e.g., [708f5d9](https://github.com/pyinstaller/pyinstaller/commit/708f5d90ff5fc91e1764ab082322ca7faf4d9c7b))? Does that work?

It does not work too...I think its somewhere in my OS that changed, but I really don't know where, and I compiled on 2 separate machines, one on Win 10 another on Win 11, and both have the same issue. Again, the 1 month old binary works on both machine.

**Reply by rokm** [reply]
Hmm... can you generate a `onedir` build of
```python
from tkinter import messagebox
messagebox.showinfo(title='xx', message='xxx')
```
and upload it somewhere so I can take a look at it?

**Reply by aliencaocao** [reply]
https://drive.google.com/file/d/14jcKJHzVBcwd1zQw5MqnB4jpBXP4-rVw/view?usp=sharing

**Reply by aliencaocao** [reply]
I just gone through my commits, and luckily I committed my binaries as well. I found out the commit that caused the change, and also what I believe the cause of the change. Something interesting changed in the manifest file. When I manually copied the manifest file from the working commit over and specified to use it, everything is working again (the pop up boxes are new). Based on what I could remember, I upgraded Python from 3.9.7 to 3.9.8. I have yet to downgrade it to try as doing so deletes all my libs, but I would be surprised if this is the reason.
![image](https://user-images.githubusercontent.com/20109683/141682830-1b3aa788-b50c-4add-a44e-1dce89e2b143.png)
Seems to me that lack of this section in the manifest directly caused the issue. I guess it has to do with Python 3.9.8

**Reply by rokm** [reply]
Yes, the lack Common Controls v.6 directive in manifest will cause the legacy controls (such as message boxes) to be used.

**Reply by aliencaocao** [reply]
Please see my updated comment. I described what I think caused the issue (upgrading python from 3.9.7 to 3.9.8). I am not sure if any changes in my code caused this to be excluded in the manifest, but I guess its due to the python upgrade, as I have no code that interact with Microsoft things.

**Reply by rokm** [reply]
It could be python upgrade, if they dropped the dependency on it. Which means we should put it in our manifest ourselves. I'll check if that is the case.

**Reply by aliencaocao** [reply]
Ok thanks. Please update me on this so that I know if in future I have to specify the manifest file myself.

**Reply by rokm** [reply]
So it looks like that with python 3.9.8, `python.exe` does not contain manifest anymore. So we do not get the `Microsoft.Windows.Common-Controls` manifest entry via dependency analysis anymore.

We do have [a fallback that activates in that case](https://github.com/pyinstaller/pyinstaller/blob/f0f63da7c7874369637f5f94225b5972ec7e419a/PyInstaller/utils/win32/winmanifest.py#L1028-L1038), but it does so only in windowed/noconsole mode. Also, it seems to not have been in use for a while, as it generates an entry that Windows does not like. So the behavior with PyInstaller `develop` and python 3.9.8 is:

* console + onedir: missing Common Controls entry in manifest, legacy common controls
* console + onefile: missing Common Controls entry in manifest, legacy common controls
* noconsole + onedir: invalid fallback Common Controls entry in manifest, leading to `The application has failed to start because its side-by-side configuration is incorrect. Please see the application event log or use the command-line sxstrace.exe tool for more detail.` error that we also saw [on CI](https://github.com/pyinstaller/pyinstaller/runs/4198998024?check_suite_focus=true#step:12:2095) in @bwoodsend's latest PR. In onedir mode, we can go back to external manifest using `--no-embed-manifest` option, which makes application start, albeit with legacy common controls (looks like Windows is more lenient with external manifest)
* noconsole + onefile: invalid fallback Common Controls entry in manifest, leading to the same incorrect side-by-side configuration error. In this case, we have no option to avoid manifest embedding.

With earlier PyInstaller versions (before #6203 and #6243 were merged to `develop`, before #6248 was merged to `v4`) the behavior should be largely the same as above, except that `noconsole` + `onedir` will start due to side-by-side manifest.

**Reply by rokm** [reply]
@aliencaocao See if #6367 fixes the problem for you.

**Reply by aliencaocao** [reply]
Yes it does.
